### ISSUE TITLE
Global costmap being cleared

### ISSUE THREAD

**Issue by lucasb-eyer** [strands-project/strands_navigation#152]
Probably I'm doing something wrong, but during backtrack behaviour the global costmap gets emptied somehow.

![goodplan](https://cloud.githubusercontent.com/assets/1476029/5147096/f8b11df6-71b2-11e4-9e19-cff7613c667f.png)

![nocost](https://cloud.githubusercontent.com/assets/1476029/5147128/4225fc18-71b3-11e4-912e-d639d4df9b78.png)

What you see in slightly darker gray is the costmap coming from the no-go map.

Output of the last time it happened:

```
[ INFO] [1416593474.709253931]: Got new plan
[ WARN] [1416593474.714531285]: DWA planner failed to produce path.
[ WARN] [1416593474.809003186]: Clearing costmap to unstuck robot (3.000000m).
[ INFO] [1416593475.209599130]: Got new plan
[ WARN] [1416593475.215077867]: DWA planner failed to produce path.
[ WARN] [1416593475.309008566]: Clearing costmap to unstuck robot (1.840000m).
[ INFO] [1416593475.709013111]: Got new plan
[ WARN] [1416593475.714326009]: DWA planner failed to produce path.
[ERROR] [1416593475.809008000]: Aborting because a valid control could not be found. Even after executing all re
covery behaviors
[INFO] [WallTime: 1416593475.822154] State machine transitioning 'NAVIGATION':'planner_failure'-->'RECOVER_NAVIG
ATION'
[INFO] [WallTime: 1416593475.823415] State machine starting in initial state 'CLEAR_COSTMAPS' with userdata: 
        ['goal', 'n_nav_fails']
[ INFO] [1416593480.954840476]: Requesting the map...
[ INFO] [1416593481.259107806]: Resizing costmap to 800 X 1400 at 0.020000 m/pix
[ INFO] [1416593481.356322228]: Received a 800 X 1400 map at 0.020000 m/pix
[ INFO] [1416593481.373878552]: Requesting the map...
[ INFO] [1416593481.576694367]: Resizing costmap to 800 X 1400 at 0.020000 m/pix
[ INFO] [1416593481.675289357]: Received a 800 X 1400 map at 0.020000 m/pix
[INFO] [WallTime: 1416593481.908846] Waiting for services...
[INFO] [WallTime: 1416593481.922603] Done
[INFO] [WallTime: 1416593481.928399] State machine transitioning 'CLEAR_COSTMAPS':'do_other_recovery'-->'BACKTRA
CK'
[INFO] [WallTime: 1416593481.933768] State machine transitioning 'BACKTRACK':'failure'-->'NAV_HELP'
```

**Comment by bfalacerda** [comment]
its the clear costmaps recovery that also clears the global costmap. I guess it should be reloaded aftewards but it isn't all the time because of the loading global costmap issue move_base has. Does this happen all the time? I'll remove the clearance from the navigation recovery, this is pretty bad...

**Comment by lucasb-eyer** [comment]
Twice for now, will watch if it happens the next times too.

**Comment by bfalacerda** [comment]
anyway, it's clearly a bad idea having it in, at least until we are sure that move_base global costmap loading is working. i'm removing it now

**Comment by lucasb-eyer** [comment]
I think it's not (only?) the clearing. If I start the system with a no-go map, _only_ the no-go map is used to create a costmap, and I get what you see in above pictures. If I start it without the costmap, the real map is correctly used to create a no-go map.

**Comment by bfalacerda** [comment]
every time you start it its not loaded?

**Comment by bfalacerda** [comment]
our no-go map is usually overlaying the original map so maybe there's an issue with the dual static map layer that we weren't aware of. From what I see your no-go only has the "no-go" obstacles right?

**Comment by lucasb-eyer** [comment]
Yeah, basically a frame around the building. It worked a few days ago as far as I can remember.

**Comment by bfalacerda** [comment]
this requires a deeper look into the issue then

**Comment by bfalacerda** [comment]
it never loads or is it only once in a while?

**Comment by lucasb-eyer** [comment]
I restarted it three times with no-go map off and three times with no-go map on. All the times no-go map was on, the global-costmap only had a halo around the no-go map; all the times no-go map was off, the global-costmap had a halo around the real map.

**Comment by bfalacerda** [comment]
all the times no-go map was off, the global-costmap had a halo around the real map.

Is this really what you wanted to say? If so I'm confused

**Comment by lucasb-eyer** [comment]
I think so. More clearly:
- with_nogo=true -> costmap only around nogo map, _not_ around real map, like screenshots above.
- with_nogo=false -> costmap around real map.

**Comment by bfalacerda** [comment]
ah ok, i misunderstood what you meant with "halo around"

I'll look into it, the easy, non system intrusive fix for now is making your nogo map the nogo areas overlaying the map

**Comment by bfalacerda** [comment]
i dont really know when/if i'll have time for this before the marathon starts...

**Comment by lucasb-eyer** [comment]
Looking into it a bit myself, otherwise I'll workaround

**Comment by bfalacerda** [comment]
you said you remember it working before. were there any changes done to our move base stuff?

**Comment by lucasb-eyer** [comment]
I really don't know when it was last working because I never actively looked for it anymore after I've last verified it worked sometimes last week.

More debugging: It's the inflation layer, which is only using the /cost_map, not the /map. Why, I don't know yet.

**Comment by bfalacerda** [comment]
http://wiki.ros.org/hydro/Migration#Navigation_-_Costmap2D

I followed this when we migrated to hydro. Not sure of why I did it like that anyomre

**Comment by bfalacerda** [comment]
the lack of inflation layer for the map probably explains why the global planner finds plans through tiny holes in the map though

**Comment by lucasb-eyer** [comment]
I looked through all tutorials and it looks like the inflation layer should just work on the fully "rendered"/combined costmap, and not on an individual layer, so something is weird. I guess I'll work around for now since time is running out.

**Comment by nilsbore** [comment]
Wasn't this exactly what @Jailander provided a fix for?

**Comment by bfalacerda** [comment]
i'll open an issue for later

**Comment by bfalacerda** [comment]
i dont recall anyone looking at costmap stuff lately but i might have missed it

**Comment by bfalacerda** [comment]
closing this, we have an open issue here:

https://github.com/strands-project/strands_movebase/issues/27
